### PR TITLE
Use CWarnings in coqdep

### LINKED ISSUE #10156: coqdep should fail when it cannot find a .v file
#### Description of the problem

Currently it emits a warning (which gets lost in Makefile output flood) and an incomplete list of dependencies (which results in the .vo not getting rebuilt).

#### Coq Version

8.9.0

### PR DESCRIPTION
**Pull request by rlepigre**
This is a revival of https://github.com/coq/coq/pull/17483, since it got closed.

Fixes https://github.com/coq/coq/issues/10156.

- [x] Added **changelog**.
- [x] ~~Added / updated **documentation**.~~

### CODE REVIEW

**Comment by rlepigre** [comment]
CC @ejgallego: this should be useful for dune builds, for which, I believe, all `coqdep` warnings should be fatal (at least by default).

In our repositories we actually wrap `coqdep` with code to detect warnings and fail if they come up, which we'd be happy to get rid of eventually.

File: tools/coqdep/lib/loadpath.ml [hunk 1]
```diff
@@ -157,11 +154,19 @@ let rec cuts recur = function
     ([],if recur then suffixes true l else [true,l]) ::
     List.map (fun (fromtail,suffixes) -> (dir::fromtail,suffixes)) (cuts true tail)
 
+let warning_ml_clash' =
+  let category = CWarnings.CoreCategories.filesystem in
+  CWarnings.create ~name:"multiple-matching-files" ~category
+    (fun (basename, suff, dir, dir') -> Pp.(
```
Comment (SkySkimmer) [inline tools/coqdep/lib/loadpath.ml:160]:
```suggestion
    Pp.(fun (basename, suff, dir, dir') ->
```

File: tools/coqdep/lib/args.ml [hunk 2]
```diff
@@ -74,9 +79,12 @@ let parse st args =
     | "-coqlib" :: [] -> usage ()
     | "-dyndep" :: dyndep :: ll -> parse { st with dyndep } ll
     | "-m" :: m :: ll -> parse { st with meta_files = st.meta_files @ [m]} ll
+    | "-w" :: w :: ll ->
+      let w = if w = "none" then w else CWarnings.get_flags() ^ "," ^ w in
+      CWarnings.set_flags w; parse st ll
```
Comment (SkySkimmer) [inline tools/coqdep/lib/args.ml:84]:
```suggestion
      CWarnings.set_flags w;
      parse st ll
```

File: tools/coqdep/coqdep.ml [hunk 3]
```diff
@@ -45,7 +47,7 @@ let coqdep () =
     if Sys.file_exists user_contrib
     then Loadpath.add_rec_dir_no_import (Loadpath.add_coqlib_known lst) user_contrib [];
     List.iter (fun s -> Loadpath.add_rec_dir_no_import (Loadpath.add_coqlib_known lst) s [])
-      (Envars.xdg_dirs ~warn:(fun x -> Warning.give "%s" x));
+      (Envars.xdg_dirs ~warn:(fun x -> Feedback.msg_warning (Pp.str x)));
```
Comment (SkySkimmer) [inline tools/coqdep/coqdep.ml:50]: There are still a few of these, do you intend to convert them to cwarnings too?

Comment (rlepigre) [reply tools/coqdep/coqdep.ml:50]: Possibly, I need to have a closer look.

Comment (rlepigre) [reply tools/coqdep/coqdep.ml]: OK, that should do.

File: tools/coqdep/coqdep.ml [hunk 4]
```diff
@@ -42,11 +44,15 @@ let coqdep () =
...
-    then Loadpath.add_rec_dir_no_import (Loadpath.add_coqlib_known lst) user_contrib [];
-    List.iter (fun s -> Loadpath.add_rec_dir_no_import (Loadpath.add_coqlib_known lst) s [])
-      (Envars.xdg_dirs ~warn:(fun x -> Warning.give "%s" x));
-    List.iter (fun s -> Loadpath.add_rec_dir_no_import (Loadpath.add_coqlib_known lst) s []) Envars.coqpath;
+    if Sys.file_exists user_contrib then
+      Loadpath.add_rec_dir_no_import (Loadpath.add_coqlib_known lst) user_contrib [];
+    let add_dir s = Loadpath.add_rec_dir_no_import (Loadpath.add_coqlib_known lst) s [] in
+    let warn =
+      let category = CWarnings.CoreCategories.filesystem in
+      CWarnings.create ~name:"cannot-locate-home-dir" ~category Pp.str
```
Comment (SkySkimmer) [inline tools/coqdep/coqdep.ml:52]: CWarnings.create should be at toplevel

File: tools/coqdep/lib/common.ml [hunk 5]
```diff
@@ -329,7 +336,10 @@ let sort st =
 let treat_coqproject st f =
   let open CoqProject_file in
   let iter_sourced f = List.iter (fun {thing} -> f thing) in
-  let warning_fn x = Warning.give "%s" x in
+  let warning_fn =
+    let category = CWarnings.CoreCategories.filesystem in
+    CWarnings.create ~name:"project-file" ~category Pp.str
```
Comment (SkySkimmer) [inline tools/coqdep/lib/common.ml:341]: CWarnings.create should be at toplevel

**Review by SkySkimmer — CHANGES_REQUESTED** [review]

File: tools/coqdep/lib/loadpath.ml [hunk 6]
```diff
@@ -157,11 +154,19 @@ let rec cuts recur = function
     ([],if recur then suffixes true l else [true,l]) ::
     List.map (fun (fromtail,suffixes) -> (dir::fromtail,suffixes)) (cuts true tail)
 
+let warning_ml_clash' =
```
Comment (SkySkimmer) [inline tools/coqdep/lib/loadpath.ml:157]: BTW usually I think we would shadow the name instead of addin `'`

Comment (rlepigre) [reply tools/coqdep/lib/loadpath.ml]: Agreed, that was not my code initially.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by rlepigre** [comment]
I added the changelog. I don't think user messages change after this MR (only potentially the way they are shown), so probably no update needed.

**Comment by rlepigre** [comment]
I also added an entry for the new argument in the `coqdep` manpage (also fixed a minor layout problem in it).

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**